### DISCUSSION TITLE
CLI flag to disable LSP

### DISCUSSION THREAD

**Discussion by dsseng** [helix-editor/helix#10226] [Ideas]
When working on large projects, you might want to simply search for something or do brief changes, not triggering LSP builds which can use a lot of RAM and battery power if you use a laptop.

Proposal: either a CLI flag or environment variable to make Helix ignore LSP server configs and not start them.

**Comment by the-mikedavis** [comment]
You can use the `-c`/`--config` file to load a configuration file at a specified path and the `editor.lsp.enable` key to enable/disable LSP. We don't want to introduce many CLI flags especially if they overlap with regular config options.

**Comment by dsseng** [comment]
Just a copy of config? Might be viable

**Comment by 7flash** [comment]
I would find it better to ask lsp validate my file after I finished typing, instead of current behavior where it's blinking with validation on every symbol while I am typing; that seems to be a default behavior in vscode too, but it's very annoying and I hope to find solution with throttle or something

**Comment by ghost** [comment]
Is `:format` supposed to be a NOP when you have `editor.lsp.enable=false` in your config? Noticed while editing rust code and being surprised `rustfmt` doesn't seem to be run.

helix 23.05 (fee8e3f7)

It works for shell scripts for which I have the recommended config from the wiki:

```toml
[[language]]
name = "bash"
indent = { tab-width = 4, unit = "    " }
formatter = { command = 'shfmt', args = ["-i", "4"] }
```

**Comment by ghost** [comment]
Since we already have `:lsp-restart` and `:lsp-stop`, how about `:lsp-start`? I disable lsp by default and would like to sometimes during a session start the server on demand. `:lsp-restart` didn't start anything, which I thought would happen, but `editor.lsp.enable=false` seems to prevent that.

**Comment by gabydd** [comment]
> Is `:format` supposed to be a NOP when you have `editor.lsp.enable=false` in your config? Noticed while editing rust code and being surprised `rustfmt` doesn't seem to be run.
> 
> helix 23.05 (fee8e3f7)
> 
> It works for shell scripts for which I have the recommended config from the wiki:
> 
> ```toml
> [[language]]
> name = "bash"
> indent = { tab-width = 4, unit = "    " }
> formatter = { command = 'shfmt', args = ["-i", "4"] }
> ```

For rust, rustfmt is actually run by rust-analyzer so the language server provides the formatting when you run ``:format`` which is why there is not formatting when lsp is disabled

**Comment by ghost** [comment]
Thank you! Fixed by adding a `rustfmt` section in `languagues.toml`. Though I would say, to avoid surprising other users who disable lsp by default, it makes sense to have the fallback built-in for use when lsp is disabled. Or you could argue someone disabling lsp could also add a `rustfmt` rule.

**Comment by willhansen** [comment]
Just ran into a use case where this feature would be convenient:

A recent change in my code is causing the LSP to take up all the RAM and freeze everything.  I want to open the editor without the LSP running for the short-term, so I can fix this problem and get everything working again.

- Calling `:lsp-stop` every time the lsp automatically restarts is inconvenient.
- Setting up a modified configuration file with this single feature disabled is also inconvenient.
- A single CLI flag would be very convenient.
